### PR TITLE
feat(nextjs): app router support

### PR DESCRIPTION
**Pull request by subzero10**
Closes: #1070

## Status
**READY**

## Description
Adds support for the new [App Router in Next.js](https://nextjs.org/docs/app/building-your-application/routing#the-app-directory).

## Todos
- [x] Implement support for app router
- [x] Make sure that source maps are correctly uploaded for the app directory
- [x] BONUS: support projects built with `.tsx` (Typescript) or `.js` files

### CODE REVIEW

**Comment by subzero10** [comment]
I also have an example project that I will push into the nextjs package with #1072.

File: packages/nextjs/src/copy-config-files.ts [hunk 1]
```diff
@@ -5,18 +5,57 @@ const fs = require('fs')
 
 const debug = process.env.HONEYBADGER_DEBUG === 'true'
 
+function usesTypescript() {
+  return fs.existsSync('tsconfig.json')
+}
+
+function usesAppRouter() {
```
Comment (BethanyBerkowitz) [inline packages/nextjs/src/copy-config-files.ts:12]: I saw in the Next documentation that they're supporting `app` and `pages` concurrently.. is that going to cause any issues ?

Comment (subzero10) [reply packages/nextjs/src/copy-config-files.ts:12]: That's a nice catch! I will have to handle that case as well 😄

**Comment by subzero10** [comment]
Error reported automatically while rendering component. Source maps applied successfully.

image

**Comment by subzero10** [comment]
Error reported automatically on the server side, during a data fetching method. However, most of the context is missing because [Next.js does not propagate server errors](https://nextjs.org/docs/app/building-your-application/routing/error-handling#handling-server-errors) to the Honeybadger React Error Boundary (since it's on the client side).

We can look into improving this with #1055.

image

**Comment by subzero10** [comment]
@BethanyBerkowitz Can I please get another round of review 🙏 ?

Thanks!

**Comment by BethanyBerkowitz** [comment]
> Error reported automatically while rendering client-side component. Source maps applied successfully.
> 
> image

This shows that the sourcemap was applied, but the file names are still minified... shouldn't they be the original file names and not minified ones?

File: packages/nextjs/src/copy-config-files.ts [hunk 2]
```diff
@@ -5,17 +5,76 @@ const fs = require('fs')
...
+  return fs.existsSync('app')
+}
+
+function getTargetPath(isAppRouter = false, isGlobalErrorComponent = false) {
+  if (!isAppRouter && isGlobalErrorComponent) {
+    throw new Error('invalid arguments: isGlobalErrorComponent can only be true when isAppRouter is true')
+  }
+
+  const extension = usesTypescript() ? 'tsx' : 'js'
```
Comment (BethanyBerkowitz) [inline packages/nextjs/src/copy-config-files.ts:25]: Why is it only either `.tsx` or `.js`? Are `.ts` and `.jsx` not allowed? (pardon my ignorance on this)

**Review by BethanyBerkowitz — COMMENTED** [review]
Code looks good to me, I left minor questions / comments. Did you want me to test this in some way?

File: packages/nextjs/src/copy-config-files.ts (hunk 2 above)
Comment (subzero10) [reply packages/nextjs/src/copy-config-files.ts:25]: It's a good question! I am following Next.js examples and these are the extensions used. So I used just these for now.

**Comment by subzero10** [comment]
> This shows that the sourcemap was applied, but the file names are still minified... shouldn't they be the original file names and not minified ones?

Are you referring to `react-dom.production.min.js`? This is a dependency, i.e. not part of the application code.
`page.tsx` is application code and appears to be shown correctly.

**Comment by subzero10** [comment]
> Did you want me to test this in some way?

I don't think it's necessary at this point, but if you want to play with Next.js, feel free to do it!

**Comment by subzero10** [comment]
@BethanyBerkowitz Can you please approve, if you are happy with my replies 😄  ?

**Pull request merged**